### PR TITLE
ENH: Speed up sparse.csgraph.dijkstra

### PR DESCRIPTION
**Pull request by windows-server-2003**
Another pull request to speed up sparse.csgraph.dijkstra, succeeding #16696
Thank you @jjerphan for reviewing the previous pull request and giving me advice

 - Replaces FibonacciHeap with std::priority_queue which makes sparse.csgraph.dijkstra faster
      - asymptotically due to the removal of a bug in FibonacciHeap implementation
      - with regard to the constant factor due to better memory usage(?)
 - Unifies the four similar dijkstra functions into one
      - Performance overhead was negligible, probably because the priority_queue is the bottleneck
 - Adds the star graph to the benchmark as the worst case for the previous version
 - Fixes a typo in the test file name 
 - A part of the test file for dijkstra is modified for better readability, nothing really changed

Benchmark(i5-10600k, ArchLinux)
```
Before:
===== ========== ============= =============
--                          format          
---------------- ---------------------------
n    min_only      random         star    
===== ========== ============= =============
30     True      178±0.1μs     125±0.9μs  
30    False      191±0.4μs     131±0.8μs  
300     True      451±0.4μs     183±0.4μs  
300    False       4.78±0ms      1.67±0ms  
900     True     2.55±0.01ms     657±2μs   
900    False      102±0.08ms   48.4±0.08ms 
===== ========== ============= =============

After:
===== ========== ============= ===========
--                         format         
---------------- -------------------------
n    min_only      random        star   
===== ========== ============= ===========
30     True      178±0.1μs     126±1μs  
30    False      192±0.6μs    134±0.8μs 
300     True      363±0.4μs    152±0.3μs 
300    False       2.46±0ms     459±1μs  
900     True     1.93±0.01ms   200±0.2μs 
900    False     48.1±0.02ms    3.78±0ms 
===== ========== ============= ===========
```

### CODE REVIEW

**Comment by larsoner** [comment]
~3x more lines removed than added, code is more readable, uses standard library functions, and improves performance. Awesome!

@fcollman I remember you working on this some years ago in #9664, do you want to take a look?

**Comment by fcollman** [comment]
I looked through it and the logic seems sound to me, and agreed I like the redesign quite a bit.  It did make me look back through the tests and wonder if we couldn't come up with some more intentionally designed use cases of graphs as they seem a bit skimpy to me.  

For example, this introduces the star format graph which I like in the benchmarks, but the results aren't validated.  Perhaps some ground truth data generated using the old implementation would be good to add to the tests and show that things still work as we desire them to.

**Comment by jjerphan** [comment]
@windows-server-2003: are you still working on this PR?

**Comment by ev-br** [comment]
A slight git glitch here --- this PR probably needs something like ` git rebase 8a99fae` (the last extra commit) or ` git rebase upstream/main`, followed by a force-push.

File: scipy/sparse/csgraph/_shortest_path.pyx [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# distutils: language = c++
+
 """
 Routines for performing shortest-path graph searches
```
Comment (fuglede) [inline scipy/sparse/csgraph/_shortest_path.pyx:5]: There are three outdated references to Fibonacci heaps in the docs (GH won't let me inline the comment but they're at lines 8, 69, 415 currently).

File: scipy/sparse/csgraph/_shortest_path.pyx [hunk 2]
```diff
@@ -618,294 +631,134 @@ def dijkstra(csgraph, directed=True, indices=None,
...
+            int[:] sources,
             DTYPE_t limit) except -1:
     cdef:
-        unsigned int Nind = dist_matrix.shape[0]
-        unsigned int N = dist_matrix.shape[1]
-        unsigned int i, k, j_source
+        unsigned int Nind = source_indices.shape[0]
+        unsigned int N = dist_matrix.shape[0]
+        unsigned int i, j_source
+        DTYPE_t next_val
```
Comment (fuglede) [inline scipy/sparse/csgraph/_shortest_path.pyx:691]: `next_val` is unused.

**Comment by tylerjereddy** [comment]
The review comments proper look pretty minor, but the comment at https://github.com/scipy/scipy/pull/17019#issuecomment-1247346395 raises some concerns about depth of testing here as we switch implementations. Given about a year of inactivity now, I'll bump the milestone again ahead of `1.13.0` branching soon-ish.

**Comment by tsery-ns** [comment]
Considering taking over. 
@windows-server-2003 is that fine by you?

**Comment by tsery-ns** [comment]
Taking over.

I have cloned @windows-server-2003's fork, rebased on current main and added a few commits to address the review. 
However, I have don't have permission to push.

Found these pages: [Allowing changes to a pull request branch created from a fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) and [Committing changes to a pull request branch created from a fork]( https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork). But I don't have the required permissions to allow changes.

Any help would be appreciated.

**Comment by j-bowhay** [comment]
@tsery-ns feel free to open a new pr with these commits

**Comment by tsery-ns** [comment]
> @tsery-ns feel free to open a new pr with these commits

PR https://github.com/scipy/scipy/pull/20717 ready for review 😄

**Pull request closed without merging**